tsconfig.json

```diff
@@ -1,8 +1,15 @@
 {
-  "extends": "@tsconfig/node16/tsconfig.json",
+  "extends": "@tsconfig/node18/tsconfig.json",
   "compilerOptions": {
+    "module": "ESNext",
+    "target": "ES2020",
+    "moduleResolution": "node",
```

Comment:
This seems to be using an old module resolution strategy. Should we update it to `node16`?

See <https://www.typescriptlang.org/tsconfig#moduleResolution>.
![image](https://github.com/MakerXStudio/node-common/assets/6102639/d390acb2-ba66-42a5-b110-30dbdbfce421)

_I also understand we only use tsc for dev time now, so maybe it doesn't matter as much._

Reply:
Great pickup thanks Mick.

Now I remember the struggles I had trying to use `NodeNext` for both `module` and `moduleResolution` setting.

Looking in the doc and the TSC compiler validation of tsconfig; it sounds as though `NodeNext` is what should be used, however with this setting:
- The rollup plugin imports stop working (I can't understand why, since Rollup exports both CJS and ES)
- Typescript imports need file extensions, which is doable but a further hassle; e.g. Jest needs transform config then to read the files with ts-jest.

![image](https://github.com/MakerXStudio/node-common/assets/5326367/6b949630-58a5-4862-89b9-b08e54782e44)

![image](https://github.com/MakerXStudio/node-common/assets/5326367/5ef06460-59de-4c21-8cc8-926f28486800)

Feel free to take a look yourself and see if you think there are other options, but I just settled on `Bundler` for module resolution, which allows extensionless typescript imports and the rollup plugin imports work under this settings.

I guess `Bundler` makes sense because in all known cases (jest + build) there is a transpiler / bundler that handles imports in a way that 'just works' ™.

What should be noted, is that we can't simply run `ts-node ./src/whatever.ts` with this configuration, ts-node and esm is a sh!tshow. Even with the ts-node esm loader, it goes: `TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for blah.ts`

I've completely lost hope in `ts-node` with ES module projects.

[`tsx`](https://github.com/privatenumber/tsx) works really well and appears to be a drop-in better replacement for `ts-node` with the caveat that it doesn't output decorator metadata. Meaning we can't use it to run projects using `typeorm`.

Thus, my current conclusion, is that right now we can consider ES module libraries, but shouldn't yet attempt to use ES module Node projects unless we know we can use [`tsx`](https://github.com/privatenumber/tsx) to run the thing. Alternatively we could have a `tsconfig.debug.ts` to use to run `.ts` files directly as CommonJS via ts-node.

Phew. It's a bit of a sh!tshow, but I think it's at least safe to output both types for libraries in most cases.

Reply:
Wow, great deep dive on this, Sam, it's much appreciated.

Bundler makes sense since tsc isn't used to produce transpiled code.